### PR TITLE
Support saving sharded checkpoints with FSDP in Trainer

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

WIP

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: src/lightning/pytorch/strategies/fsdp.py [hunk 1]
```diff
@@ -168,23 +179,12 @@ def __init__(
             activation_checkpointing, activation_checkpointing_policy
         )
 
-    def lightning_module_state_dict(self) -> Dict[str, Any]:
```
Comment (awaelchli) [inline src/lightning/pytorch/strategies/fsdp.py:171]: Moved this down to the other methods

**Review by Borda — APPROVED** [review]

File: src/lightning/pytorch/strategies/fsdp.py [hunk 2]
```diff
@@ -440,6 +440,43 @@ def register_strategies(cls, strategy_registry: _StrategyRegistry) -> None:
         )
         cls._registered_strategies.append("fsdp_cpu_offload")
 
+    def lightning_module_state_dict(self) -> Dict[str, Any]:
+        assert self.model is not None
+        if self._state_dict_type == "sharded":
+            state_dict_ctx = _get_sharded_state_dict_context(self.model)
+        elif self._state_dict_type == "full":
+            state_dict_ctx = _get_full_state_dict_context(self.model, offload_to_cpu=(self.world_size > 1))
```
Comment (carmocca) [inline src/lightning/pytorch/strategies/fsdp.py:448]: Does Fabric also need this `world_size > 1` logic? I don't find it there

**Review by carmocca — APPROVED** [review]

File: src/lightning/pytorch/strategies/fsdp.py (hunk 2 above)
Comment (awaelchli) [reply src/lightning/pytorch/strategies/fsdp.py:448]: I couldn't replicate an issue with world_size=1 for FSDP and offload_to_cpu=True so I'm not sure. Leaving it as is for now.

Comment (carmocca) [reply src/lightning/pytorch/strategies/fsdp.py:448]: Thats strange. What's the issue? It would be good to describe it here for posterity, even if left as is for now

File: src/lightning/pytorch/strategies/fsdp.py [hunk 3]
```diff
@@ -168,23 +179,12 @@ def __init__(
...
-
-        """
-        from torch.distributed.fsdp import FullStateDictConfig, FullyShardedDataParallel, StateDictType
-
-        assert self.model is not None
-
-        with FullyShardedDataParallel.state_dict_type(
-            module=self.model,
-            state_dict_type=StateDictType.FULL_STATE_DICT,
-            state_dict_config=FullStateDictConfig(offload_to_cpu=(self.world_size > 1), rank0_only=True),
```
Comment (awaelchli) [inline src/lightning/pytorch/strategies/fsdp.py:185]: @carmocca See here, the code was here before, I'm merely just refactoring it so that it uses the existing logic from Fabric. The condition `offload_to_cpu=(self.world_size > 1)` is not new to this PR, it was there before.

Comment (carmocca) [reply src/lightning/pytorch/strategies/fsdp.py:185]: Ah I missed this. Thanks for the context

File: src/lightning/pytorch/strategies/fsdp.py (hunk 2 above)
Comment (awaelchli) [reply src/lightning/pytorch/strategies/fsdp.py:448]: Ah ok, it is only an issue in torch 2.0. I'll make sure we standardize this across both PyTorch and Fabric in #18371.
